waspc/data/Generator/templates/server/src/auth/providers/oauth/user.ts

```diff
@@ -78,12 +84,36 @@ async function getAuthIdFromProviderDetails({
   })
 
   if (existingAuthIdentity) {
-    return existingAuthIdentity.{= authFieldOnAuthIdentityEntityName =}.id
+    const authId = existingAuthIdentity.{= authFieldOnAuthIdentityEntityName =}.id
+
+    // NOTE: We are calling login hooks here even though we didn't log in the user yet.
+    // We are doing it here because we have access to the OAuth tokens and we can pass them to the hooks.
+    // This isn't a big deal because the next step of the OAuth flow happens immediately after this function
+    // and the user is redirected to the client with the one-time code which is then used to create the session.
+    // The downside of this approach is that we can't provide the session to the login hooks, but this is
+    // an okay trade-off for now.
+    await onBeforeLoginHook({ req, providerId })
```

Comment:
I'm not sure I understand this.

So, both hooks are called before the login. Shouldn't this comment then be above the `onAfterLooginHook` call (since this is the hook that's called when it's supposed to be called, while that one isn't)?

If that's the case, can you explain (in the comment), why having access to the OAuth token is more important than calling the hook at the right time.

Reply:
Both of these hooks are called differently than for email and username providers, so the comment still makes sense above `onBeforeLoginHook`. For email and username, we call `onBeforeLoginHook` right before the `createSession` call and the `onAfterLoginHook` right after that. 

For OAuth providers, we call them both right before we redirect to the client and then the session is created when the client exchanges the one time code with the server for the session. 

The OAuth tokens are _only available here_ and never later, and these tokens are very valuable for our users because they enable various API integrations e.g. using Google Sheets with their users' credentials. We could temporarily store the token and then make it available in the one time code <-> session exchange handler, but that felt needlessly complicated. The session ID is not that important and this a good trade off. 

I'll write these details in the comment above 👍